tls, crypto: add ALPN Support

shigeki opened the pull request:
ALPN is added to tls according to RFC7301, which supersedes NPN. 
When the server receives both NPN and ALPN extensions from the client, ALPN takes precedence over NPN and the server does not send NPN extension to the client. alpnProtocol in TLSSocket always returns false when no selected protocol exists by ALPN.
In https server, http/1.1 token is always set when no options.ALPNProtocols exists.

Here exist all 16x3 test cases of combination of  ALPN and NPN in a server and a client. Tests shows that there are some inconsistent returns in NPN between the server and the client but they are not changed for compatibility.

NPN in Chrome will be deprecated in early 2016 as in http://blog.chromium.org/2015/02/hello-http2-goodbye-spdy-http-is_9.html so it is better to include ALPN support in Node 4.0.

R= @nodejs/crypto

indutny commented on src/node_crypto.cc:
```diff
@@ -1154,6 +1163,12 @@ void SSLWrap<Base>::AddMethods(Environment* env, Handle<FunctionTemplate> t) {
 #endif
 
+#ifdef TLSEXT_TYPE_application_layer_protocol_negotiation
+  NODE_SET_PROTOTYPE_METHOD(t, "getALPNNegotiatedProtocol",
```
What if we will always set and declare them, but just leave their implementations empty in case of absent ALPN support?

bnoordhuis replied:
@shigeki I guess you can leave the #ifdef out.  I remember from the last review that it looked like it wouldn't compile with ALPN disabled but maybe I saw that wrong.

shigeki replied:
I removed #ifdef here and fixed to have declaration to avoid compile errors and also changed `NODE_SET_PROTOTYPE_METHOD` macro into ` env->SetProtoMethod`.

indutny commented on src/node_crypto.cc:
```diff
@@ -1972,6 +1987,89 @@ void SSLWrap<Base>::SetNPNProtocols(const FunctionCallbackInfo<Value>& args) {
+      reinterpret_cast<const unsigned char*>(Buffer::Data(obj));
+  size_t len = Buffer::Length(obj);
+
+  int status = SSL_select_next_proto(
```
I didn't know this existed!

indutny commented on src/node_crypto.cc:
```diff
@@ -1972,6 +1987,89 @@ void SSLWrap<Base>::SetNPNProtocols(const FunctionCallbackInfo<Value>& args) {
+      int r = SSL_set_alpn_protos(w->ssl_, alpn_protos, alpn_protos_len);
+      assert(r == 0);
+    } else {
+      w->alpn_protos_.Reset(args.GetIsolate(), args[0].As<Object>());
```
What do you think about allocating and copying the data instead? I think I'm mostly against using Persistent for this reasons now.

indutny commented:
@shigeki this test is awesome, thanks! Some nits, otherwise looking good.

shigeki commented:
@indutny Thanks very much for your quick review. I fixed this according your comments and I also found that Test12 was wrong to be duplicated from Test11 and fixed it to the test case of Server: NPN, Client: Nothing.
CI of https://jenkins-iojs.nodesource.com/job/node-test-commit-linux/315/ is all green. 
I also made h2 server tests connected from Chrome(boringSSL) and Firefox(nss) with only ALPN and ALPN/NPN and it works fine.  Unfortunately I don't have MS Edge so did not test it yet.

Can I land this?

indutny commented on lib/_tls_wrap.js:
```diff
@@ -951,9 +962,10 @@ exports.connect = function(/* [port, host], options, cb */) {
                  (options.socket && options.socket._host) ||
                  'localhost',
-      NPN = {},
+      NPN = {}, ALPN = {},
```
Please put it on the next line.

shigeki replied:
Okay, I will fix this.

indutny commented on src/node_crypto.cc:
```diff
@@ -1972,6 +1985,91 @@ void SSLWrap<Base>::SetNPNProtocols(const FunctionCallbackInfo<Value>& args) {
 
+#ifdef TLSEXT_TYPE_application_layer_protocol_negotiation
+typedef struct tlsextalpnctx_st {
+  unsigned char *data;
```
`unsigned char* data`, please. (/me wish linter will detect this one day)

shigeki replied:
fixed.

indutny commented on src/node_crypto.cc:
```diff
@@ -1972,6 +1985,91 @@ void SSLWrap<Base>::SetNPNProtocols(const FunctionCallbackInfo<Value>& args) {
+      int r = SSL_set_alpn_protos(w->ssl_, alpn_protos, alpn_protos_len);
+      assert(r == 0);
+    } else {
+      w->alpn_protos_ =
```
You'll need to keep the reference in JS land, otherwise "bad things will happen" when the buffer will be GC-ed. I'd suggest storing the reference on the context itself.

shigeki replied:
Okay, I will fix this.

shigeki replied:
> What do you think about allocating and copying the data instead? I think I'm mostly against using Persistent for this reasons now.

Let me confirm this. Is this what you mean or another way to avoid using Persistent? Maybe I lost the point. Thanks.
```diff
@@ -2063,7 +2063,10 @@ void SSLWrap<Base>::SetALPNProtocols(
       int r = SSL_set_alpn_protos(w->ssl_, alpn_protos, alpn_protos_len);
       assert(r == 0);
     } else {
-      w->alpn_protos_.Reset(args.GetIsolate(), args[0].As<Object>());
+      w->alpn_protos_.Reset(args.GetIsolate(), Buffer::Copy(
+          w->env(),
+          reinterpret_cast<const char*>(Buffer::Data(args[0])),
+          Buffer::Length(args[0])).ToLocalChecked());
       // Server should select ALPN protocol from list of advertised by client
       SSL_CTX_set_alpn_select_cb(w->ssl_->ctx, SelectALPNCallback, nullptr);
     }
````

indutny replied:
Ah, I meant doing this in js:

```javascript
secureContext._alpnBuffer = alpnBuffer;
secureContext.setALPNProtocols(alpnBuffer);
```

This way you can just use the `Buffer::Data` and be sure that it won't be freed while the `secureContext` is alive itself.

shigeki replied:
Thanks very much, I've got it. Is this good for it?
```diff
diff --git a/lib/_tls_legacy.js b/lib/_tls_legacy.js
index 2f26c90..4686e98 100644
--- a/lib/_tls_legacy.js
+++ b/lib/_tls_legacy.js
@@ -727,7 +727,9 @@ function SecurePair(context, isServer, requestCert, rejectUnauthorized,
   }

   if (process.features.tls_alpn && options.ALPNProtocols) {
-    this.ssl.setALPNrotocols(options.ALPNProtocols);
+    // keep reference in secureContext not to be GC-ed
+    this.ssl._secureContext.alpnBuffer = options.ALPNProtocols;
+    this.ssl.setALPNrotocols(this.ssl._secureContext.alpnBuffer);
     this.alpnProtocol = null;
   }

diff --git a/lib/_tls_wrap.js b/lib/_tls_wrap.js
index c7fc5b3..e1da127 100644
--- a/lib/_tls_wrap.js
+++ b/lib/_tls_wrap.js
@@ -451,8 +451,11 @@ TLSSocket.prototype._init = function(socket, wrap) {
   if (process.features.tls_npn && options.NPNProtocols)
     ssl.setNPNProtocols(options.NPNProtocols);

-  if (process.features.tls_alpn && options.ALPNProtocols)
-    ssl.setALPNProtocols(options.ALPNProtocols);
+  if (process.features.tls_alpn && options.ALPNProtocols) {
+    // keep reference in secureContext not to be GC-ed
+    ssl._secureContext.alpnBuffer = options.ALPNProtocols;
+    ssl.setALPNProtocols(ssl._secureContext.alpnBuffer);
+  }

   if (options.handshakeTimeout > 0)
     this.setTimeout(options.handshakeTimeout, this._handleTimeout);
````

shigeki commented:
The milestone was deleted not to block 4.0 release. 

@indutny Sorry, I am in a vacation from today and will work on this on the next Saturday.

argon commented:
What's the status of this? I may have a future dependency on http2 in a project and no ALPN may be a blocker.

shigeki commented:
In Chrome, "Disable HTTP/2 over NPN (with OpenSSL)" https://codereview.chromium.org/1387363004 was landed today. It is better to include this in 5.0

I've rebased this against the latest master and CI is  https://ci.nodejs.org/job/node-test-commit/912/ . Something wrong with tests environments is on freebsd and windows and plinux test failures are tick-processor.js and test-debug-args.js that are nothing to do with this PR. Otherwise tests are green.

@indutny Your comments are included in https://github.com/shigeki/node/commit/9625a828a9c5c739e163c2200726327b8bf26a03. LGTM?

shigeki commented:
@argon Sorry for being late. I've just update this. Thanks for patience.

argon commented:
Thanks for the update!

bnoordhuis commented on doc/api/tls.markdown:
```diff
@@ -66,14 +66,15 @@ and tap `R<CR>` (that's the letter `R` followed by a carriage return) a few
 <!-- type=misc -->
 
-NPN (Next Protocol Negotiation) and SNI (Server Name Indication) are TLS
+ALPN(Application-Layer Protocol Negotiation Extension), NPN (Next
```
Space before (.

bnoordhuis commented on doc/api/tls.markdown:
```diff
@@ -249,6 +250,12 @@ automatically set as a listener for the [secureConnection][] event.  The
+    protocols. (Protocols should be ordered by their priority). When
+    the server receives both NPN and ALPN extensions from the client,
+    ALPN takes precedence over NPN and the server does not send NPN
+    extension to the client.
```
s/extension/extensions/?

shigeki replied:
In this case, only one extension of NPN is not returned so I think it is singular.

bnoordhuis replied:
Then it should be 'an NPN extension' (or possibly 'the' - reads a little awkward though.)

shigeki replied:
Thanks, I fixed it. English article is very hard for me ;-0

bnoordhuis commented on doc/api/tls.markdown:
```diff
@@ -376,6 +383,12 @@ Creates a new client connection to the given `port` and `host` (old API) or
     usually be much simpler: `['hello', 'world']`.)
 
+  - `ALPNProtocols`: An array of strings or `Buffer`s containing
+    supported ALPN protocols. `Buffer`s should have following format:
```
'the following'

bnoordhuis commented on doc/api/tls.markdown:
```diff
@@ -376,6 +383,12 @@ Creates a new client connection to the given `port` and `host` (old API) or
+  - `ALPNProtocols`: An array of strings or `Buffer`s containing
+    supported ALPN protocols. `Buffer`s should have following format:
+    `0x05hello0x05world`, where first byte is next protocol name's
+    length. (Passing array should usually be much simpler:
```
'the first byte' and 'the next protocol' and 'an array'.

bnoordhuis commented on doc/api/tls.markdown:
```diff
@@ -571,7 +586,13 @@ server. If `socket.authorized` is false, then
-`socket.npnProtocol` is a string containing selected NPN protocol.
+
+`socket.npnProtocol` is a string containing selected NPN protocol and
+`socket.alpnProtocol` is a string containing selected ALPN, When both
```
'the selected' (2x)

bnoordhuis commented on doc/api/tls.markdown:
```diff
@@ -744,8 +765,9 @@ The listener will be called no matter if the server's certificate was
-`tlsSocket.npnProtocol` for negotiated protocol.
+`tlsSocket.authorizationError`. Also if NPN or ALPN was used - you can
+check `tlsSocket.npnProtocol` or `tlsSocket.alpnProtocol` for
+negotiated protocol.
```
'the negotiated protocol'

bnoordhuis commented on lib/_tls_wrap.js:
```diff
@@ -453,6 +454,12 @@ TLSSocket.prototype._init = function(socket, wrap) {
 
+  if (process.features.tls_alpn && options.ALPNProtocols) {
+    // keep reference in secureContext not to be GC-ed
+    ssl._secureContext.alpnBuffer = options.ALPNProtocols;
```
This may be dangerous if `options.ALPNProtocols` is modified by the user ex post facto?

shigeki replied:
In the case that `options.ALPNProtocols` is Array, it is converted to buffer of TLS vector format with `tls.convertALPN`. I made a fix in case that it's type is Buffer.

bnoordhuis replied:
@indutny I think you suggested this approach?  Can you comment on whether this is what you had in mind?

bnoordhuis replied:
Also, why doesn't the code for NPN keep a reference?  It's implemented pretty much the same way, isn't it?

shigeki replied:
This change comes from the old comment from @indutny  that

````
> +      w->alpn_protos_.Reset(args.GetIsolate(), args[0].As<Object>());
  What do you think about allocating and copying the data instead? I think I'm mostly against using Persistent for this reasons now.
````


I did not think of the reasons so much. I'm 0 for this change but I think we had better to have consistency between ALPN and NPN implementations.

bnoordhuis commented on src/node_crypto.cc:
```diff
@@ -2010,6 +2023,91 @@ void SSLWrap<Base>::SetNPNProtocols(const FunctionCallbackInfo<Value>& args) {
+  HandleScope handle_scope(env->isolate());
+  Context::Scope context_scope(env->context());
+
+  int status = SSL_select_next_proto((unsigned char**) out, outlen,
```
const_cast?

shigeki replied:
The openssl api is not const as.
> int SSL_select_next_proto(unsigned char **out, unsigned char *outlen .. )

bnoordhuis replied:
I realize that.  My nit is about using a C-style cast instead of a C++ const_cast.

shigeki replied:
Oops, I didn't think of it. Sorry, I fixed it in a55984128754ebcdf85b5bc49d317076141e498e and will make squash later.

bnoordhuis commented on src/node_crypto.cc:
```diff
@@ -2010,6 +2023,91 @@ void SSLWrap<Base>::SetNPNProtocols(const FunctionCallbackInfo<Value>& args) {
+      // According to 3.2. Protocol Selection of RFC7301,
+      // fatal no_application_protocol alert shall be sent
+      // but current openssl does not support it yet.
+      // See https://rt.openssl.org/Ticket/Display.html?id=3463.
```
Is that ticket embargoed?  It wants me to sign in.

bnoordhuis commented on src/node_crypto.cc:
```diff
@@ -2010,6 +2023,91 @@ void SSLWrap<Base>::SetNPNProtocols(const FunctionCallbackInfo<Value>& args) {
+  if (args.Length() < 1 || !Buffer::HasInstance(args[0]))
+    return w->env()->ThrowTypeError("Must give a Buffer as first argument");
+
+    if (w->is_client()) {
```
The indentation is wrong here.

bnoordhuis commented on src/node_crypto.cc:
```diff
@@ -2010,6 +2023,91 @@ void SSLWrap<Base>::SetNPNProtocols(const FunctionCallbackInfo<Value>& args) {
+          reinterpret_cast<const unsigned char*>(Buffer::Data(args[0]));
+      unsigned alpn_protos_len = Buffer::Length(args[0]);
+      int r = SSL_set_alpn_protos(w->ssl_, alpn_protos, alpn_protos_len);
+      assert(r == 0);
```
Can you use CHECK_EQ here?

bnoordhuis commented on src/node_crypto.cc:
```diff
@@ -1148,6 +1157,10 @@ void SSLWrap<Base>::AddMethods(Environment* env, Local<FunctionTemplate> t) {
 
+  NODE_SET_PROTOTYPE_METHOD(t, "getALPNNegotiatedProtocol",
+                            GetALPNNegotiatedProto);
+  NODE_SET_PROTOTYPE_METHOD(t, "setALPNProtocols", SetALPNProtocols);
```
Shouldn't this be wrapped in an `#ifdef TLSEXT_TYPE_application_layer_protocol_negotiation`?

bnoordhuis commented on src/node_crypto.h:
```diff
@@ -298,6 +317,12 @@ class SSLWrap {
+#ifdef TLSEXT_TYPE_application_layer_protocol_negotiation
+  const unsigned char* alpn_protos_;
+  unsigned alpn_protos_len_;
+  const unsigned char* selected_alpn_proto_;
```
I think you should set these to nullptr / zero here or in the constructor.  I'd do it here.

bnoordhuis commented on test/parallel/test-tls-alpn-server-client.js:
```diff
@@ -0,0 +1,540 @@
+'use strict';
+var common = require('../common');
```
Can you use `const` for import statements here and below?

shigeki commented:
@bnoordhuis Thanks for reviewing. According to your comments, I made fixes in 5c7c6082b3859dfd6885a8c7549f710e788234aa, otherwise I  put comment replies. PTAL.

shigeki commented:
@bnoordhuis Thanks for your comments. Additional fixes were made in a55984128754ebcdf85b5bc49d317076141e498e and will be squashed later.

indutny commented:
LGTM

shigeki commented:
@indutny Thanks.
@bnoordhuis Could you give me a final LGTM before landing?

I tried to run CI but it has something wrong in building addon tests now.

jasnell commented:
CI should be fixed now. New run kicked off... https://ci.nodejs.org/job/node-test-pull-request/575/

shigeki commented:
@jasnell Thanks! I appreciate your help.

jasnell commented:
CI is looking good. I haven't done a detailed review yet tho.

shigeki commented:
@bnoordhuis Ping. Could you give me a final LGTM?

bnoordhuis commented on doc/api/tls.markdown:
```diff
@@ -571,7 +587,13 @@ server. If `socket.authorized` is false, then
-`socket.npnProtocol` is a string containing selected NPN protocol.
+
+`socket.npnProtocol` is a string containing the selected NPN protocol
+and `socket.alpnProtocol` is a string containing the selected ALPN,
```
'the selected ALPN' - did a word fall off here?  'protocol' maybe?

shigeki replied:
The word of `protocol` was missed. Fixed.

bnoordhuis commented on doc/api/tls.markdown:
```diff
@@ -744,8 +766,9 @@ The listener will be called no matter if the server's certificate was
 If `tlsSocket.authorized === false` then the error can be found in
-`tlsSocket.authorizationError`. Also if NPN was used - you can check
-`tlsSocket.npnProtocol` for negotiated protocol.
+`tlsSocket.authorizationError`. Also if NPN or ALPN was used - you can
```
Maybe write this as `if ALPN or NPN was used` to hint at the precedence order.

shigeki replied:
Fixed.

bnoordhuis commented on doc/api/tls.markdown:
```diff
@@ -744,8 +766,9 @@ The listener will be called no matter if the server's certificate was
-`tlsSocket.authorizationError`. Also if NPN was used - you can check
-`tlsSocket.npnProtocol` for negotiated protocol.
+`tlsSocket.authorizationError`. Also if NPN or ALPN was used - you can
+check `tlsSocket.npnProtocol` or `tlsSocket.alpnProtocol` for the
```
Likewise here.

shigeki replied:
Fixed.

bnoordhuis commented on lib/tls.js:
```diff
@@ -33,27 +33,42 @@ exports.getCiphers = function() {
-    }, 0);
-
-    NPNProtocols = buff;
+function convertProtocols(Protocols) {
```
Maybe do `s/Protocols/protocols/` in this function

shigeki replied:
Fixed all first letters of argument name of `Protocols` here in lower-case.

bnoordhuis commented on src/node_crypto.cc:
```diff
@@ -2010,6 +2025,91 @@ void SSLWrap<Base>::SetNPNProtocols(const FunctionCallbackInfo<Value>& args) {
+    CHECK_EQ(r, 0);
+  } else {
+    w->alpn_protos_ =
+        reinterpret_cast<const unsigned char*>(Buffer::Data(args[0]));
```
I'd keep the reference to the buffer in a `v8::Persistent<v8::Object>`.  Doing it in JS land is not obvious and easy to break.

shigeki replied:
I'm 0 for this. I'd like to have @indutny 's reply.

indutny replied:
@bnoordhuis perhaps a hidden property instead?

bnoordhuis replied:
That seems reasonable.

shigeki replied:
It seems that `JSObject::SetHiddenProperty` is not a public API so is it done by using `Object.defineProperty` in JS land like following diff? Or is it something else?
```diff
diff --git a/lib/_tls_legacy.js b/lib/_tls_legacy.js
index 8c079e3..b3f6385 100644
--- a/lib/_tls_legacy.js
+++ b/lib/_tls_legacy.js
@@ -727,8 +727,14 @@ function SecurePair(context, isServer, requestCert, rejectUnauthorized,
   }

   if (process.features.tls_alpn && options.ALPNProtocols) {
-    // keep reference in secureContext not to be GC-ed
-    this.ssl._secureContext.alpnBuffer = options.ALPNProtocols;
+    // Keep reference in a hidden property of secureContext not to be
+    // modified by user and GC-ed.
+    Object.defineProperty(this.ssl._secureContext, 'alpnBuffer', {
+      enumerable: false,
+      configurable: false,
+      writable: false,
+      value: options.ALPNProtocols
+    });
     this.ssl.setALPNrotocols(this.ssl._secureContext.alpnBuffer);
     this.alpnProtocol = null;
   }
diff --git a/lib/_tls_wrap.js b/lib/_tls_wrap.js
index d918656..f2438cf 100644
--- a/lib/_tls_wrap.js
+++ b/lib/_tls_wrap.js
@@ -455,8 +455,14 @@ TLSSocket.prototype._init = function(socket, wrap) {
     ssl.setNPNProtocols(options.NPNProtocols);

   if (process.features.tls_alpn && options.ALPNProtocols) {
-    // keep reference in secureContext not to be GC-ed
-    ssl._secureContext.alpnBuffer = options.ALPNProtocols;
+    // Keep reference in a hidden property of secureContext not to be
+    // modified by user and GC-ed.
+    Object.defineProperty(ssl._secureContext, 'alpnBuffer', {
+      enumerable: false,
+      configurable: false,
+      writable: false,
+      value: options.ALPNProtocols
+    });
     ssl.setALPNProtocols(ssl._secureContext.alpnBuffer);
   }
````

bnoordhuis replied:
Fedor was referring to `v8::Object::SetHiddenValue()`.  `Object.defineProperty()` works just as well but in that case you should add a comment explaining that a reference is maintained in JS land.

bnoordhuis commented on src/node_crypto.h:
```diff
@@ -5,7 +5,8 @@
+#if defined(OPENSSL_NPN_NEGOTIATED) || \
+  defined(TLSEXT_TYPE_application_layer_protocol_negotiation)
 #include "node_buffer.h"
 #endif
```
I'd be alright with just dropping the #if guard here.

shigeki replied:
Yes, Buffer is used out of ALPN/NPN so that we no longer have its guard. Fixed.

bnoordhuis commented on src/node_crypto.h:
```diff
@@ -259,6 +267,17 @@ class SSLWrap {
 #endif  // OPENSSL_NPN_NEGOTIATED
+#ifdef TLSEXT_TYPE_application_layer_protocol_negotiation
+  static void GetALPNNegotiatedProto(
+    const v8::FunctionCallbackInfo<v8::Value>& args);
```
I think this is where I thought it wouldn't compile with ALPN disabled (and it wouldn't, the declaration is inside the #ifdef block.)

shigeki replied:
Yes, I've tested and this causes compile error. #ifdef  is removed in here and wrong wrapping indents are also fixed. #ifdef in SetMethod is also removed for the functions to return undefined.

shigeki commented:
Thaks for reviewing, Ben. The fixes were made in 9ac444fa59557026a9d57a824ec9950f1d86856e except of 
`w->alpn_protos_` comment.  I'd like to have an opinion from @indutny for this change.

shigeki commented:
@bnoordhuis Thanks for I missed to find `HiddenValue()`. I fixed this so as to put `alpnBuffer` as a hidden value into `w->object()`.   Is ff92f05dedc3b52f805b5fa0b44c6c7ae21a7252 a right answer?

bnoordhuis commented:
Yes.  I suggest you run the CI one more time but LGTM.

shigeki commented:
@bnoordhuis Thanks. I've just added a new commit to fix NPN implementation to be consistent with ALPN one to use hidden values and remove pesistents. Could you also take a look at 17e560ab54f59a9dcb9d11b817d8e9371ca4480c ? 
CI is https://ci.nodejs.org/job/node-test-commit/975/ and all results are fine except win-vs2013 that is failed due to ssh-agent error.

bnoordhuis commented:
> Could you also take a look at 17e560a ? 

LGTM

shigeki commented:
@bnoordhuis @indutny Thanks for reviewing. Landed in 856d1aaeec0cba02cb939966aafd356fa3af5342 and d9e7a5a402dc05d9b0a85ec8cc19f26c3677728a on 5.x branch.

bnoordhuis commented:
@shigeki You probably missed the memo but all commits should go onto the master branch.  The release managers will take care of cherry-picking it to the release branches.

shigeki commented:
@bnoordhuis @rvagg @Fishrock123  Oops, Sorry.  Should I force to push or leave it?

bnoordhuis commented:
I'd skip the 10 minute rule and force-push away the commits on the v5.x branch and land them on master.

shigeki commented:
Yes, I've just to force push v5.x after 12mins.  2mins delay, sorry.

shigeki commented:
Landed 802a2e79e1adb22542ba12fba5e331e94277272d and 7eee37257fe708ec23e8be292047320491ac8361 on master. Sorry for confusion.

MylesBorins commented:
@nodejs/crypto I attempted a backport of this but all the tests blew up. This is touching far too much code for me to comfortable backport. Would someone be able to get a one in next week?

shigeki commented:
Okay, I can work it next week.

shigeki commented:
@MylesBorins I've just submitted the PR of #10831 .
